builder/swig/csharp.bzl

```diff
@@ -0,0 +1,241 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+# TODO: Generalise and move the same functions (used by java, python, csharp)
+# to a base file, add reusage where possible.
+
+load("@rules_dotnet//dotnet:defs.bzl", "csharp_library")
+load("@rules_dotnet//dotnet/private:providers.bzl", "DotnetAssemblyCompileInfo", "DotnetAssemblyRuntimeInfo")
+
+
+def _swig_csharp_wrapper_impl(ctx):
+    module_name = getattr(ctx.attr, "class_name", ctx.attr.name)
+
+    wrap_csharp_name = "{}.cs".format(module_name)
+    wrap_csharp = ctx.actions.declare_file(wrap_csharp_name)
+
+    args = ctx.attr.extra_args + [
+        "-csharp",
+        "-namespace", ctx.attr.namespace,
+        "-outfile", wrap_csharp_name,
+        ctx.file.interface.path,
+    ]
+
+    if ctx.attr.enable_cxx:
+        wrap_cxx = ctx.actions.declare_file("{}_wrap.cxx".format(module_name))
+        directors_header = ctx.actions.declare_file("{}_wrap.h".format(module_name))
+        args = ["-c++", "-o", wrap_cxx.path, "-oh", directors_header.path] + args
+        swig_headers = [directors_header]
+    else:
+        wrap_cxx = ctx.actions.declare_file("{}_wrap.c".format(module_name))
+        args = ["-o", wrap_cxx.path] + args
+        swig_headers = []
+    
+    for h in ctx.attr.lib[CcInfo].compilation_context.headers.to_list():
+        args = ["-I" + h.dirname] + args
+
+    ctx.actions.run(
+        inputs = depset(
+            [ctx.file.interface] + ctx.files.includes,
+            transitive = [
+                ctx.attr.lib[CcInfo].compilation_context.headers,
+                ctx.attr._swig.data_runfiles.files,
+        ]),
+        outputs = [wrap_cxx, wrap_csharp] + swig_headers,
+        executable = ctx.file._swig,
+        arguments = args,
+    )
+
+    lib_compilation_context = ctx.attr.lib[CcInfo].compilation_context
+    compilation_context = cc_common.create_compilation_context(
+        headers = depset(swig_headers, transitive = [lib_compilation_context.headers]),
+        defines = lib_compilation_context.defines,
+        framework_includes = lib_compilation_context.framework_includes,
+        includes = depset(
+            [h.dirname for h in ctx.attr.lib[CcInfo].compilation_context.headers.to_list()],
+            transitive = [lib_compilation_context.includes],
+        ),
+        local_defines = lib_compilation_context.local_defines,
+        quote_includes = lib_compilation_context.quote_includes,
+        system_includes = depset(
+            [],
+            transitive = [lib_compilation_context.system_includes],
+        )
+    )
+
+    return [
+        DefaultInfo(files = depset([wrap_cxx, wrap_csharp])),
+        OutputGroupInfo(
```

Comment:
This is a useful technique to save separate sets of files for future usage. I had troubles with separating C# and CXX files from each other to compile both source and test libraries, and it was a perfect solution!